src/limbo/model/sparse_gp.hpp

```diff
@@ -0,0 +1,208 @@
+//|   - Vaios Papaspyros (b.papaspyros@gmail.com)
+//|   - Roberto Rama (bertoski@gmail.com)
+//|
+//| This software is a computer library whose purpose is to optimize continuous,
+//| black-box functions. It mainly implements Gaussian processes and Bayesian
+//| optimization.
+//| Main repository: http://github.com/resibots/limbo
+//| Documentation: http://www.resibots.eu/limbo
+//|
+//| This software is governed by the CeCILL-C license under French law and
+//| abiding by the rules of distribution of free software.  You can  use,
+//| modify and/ or redistribute the software under the terms of the CeCILL-C
+//| license as circulated by CEA, CNRS and INRIA at the following URL
+//| "http://www.cecill.info".
+//|
+//| As a counterpart to the access to the source code and  rights to copy,
+//| modify and redistribute granted by the license, users are provided only
+//| with a limited warranty  and the software's author,  the holder of the
+//| economic rights,  and the successive licensors  have only  limited
+//| liability.
+//|
+//| In this respect, the user's attention is drawn to the risks associated
+//| with loading,  using,  modifying and/or developing or reproducing the
+//| software by the user in light of its specific status of free software,
+//| that may mean  that it is complicated to manipulate,  and  that  also
+//| therefore means  that it is reserved for developers  and  experienced
+//| professionals having in-depth computer knowledge. Users are therefore
+//| encouraged to load and test the software's suitability as regards their
+//| requirements in conditions enabling the security of their systems and/or
+//| data to be ensured and,  more generally, to use and operate it in the
+//| same conditions as regards security.
+//|
+//| The fact that you are presently reading this means that you have had
+//| knowledge of the CeCILL-C license and that you accept its terms.
+//|
+#ifndef LIMBO_MODEL_SPARSE_GP_HPP
+#define LIMBO_MODEL_SPARSE_GP_HPP
+
+#include <algorithm>
+#include <mutex>
+
+#include <limbo/model/gp.hpp>
+#include <limbo/tools/parallel.hpp>
+
+namespace limbo {
+    namespace defaults {
+        struct model_sparse_gp {
+            BO_PARAM(int, max_points, 200);
+        };
+    } // namespace defaults
+
+    namespace model {
+        /// @ingroup model
+        /// A sparse Gaussian process model.
+        /// It is parametrized by:
+        /// - a kernel function
+        /// - a mean function
+        /// - [optional] an optimizer for the hyper-parameters
+        /// A sparsification based on the density of points is performed
+        /// until a desired number of points is reached
+        template <typename Params, typename KernelFunction = kernel::MaternFiveHalves<Params>, typename MeanFunction = mean::Data<Params>, typename HyperParamsOptimizer = gp::NoLFOpt<Params>>
+        class SparseGP : public GP<Params, KernelFunction, MeanFunction, HyperParamsOptimizer> {
```

Comment:
should we name this differently? this seems a bit generic (there are many 'sparseGP'). Maybe DensitySparseGP? or something elese

Reply:
SparsifiedGP ?